### PR TITLE
feat: add sandbox logic

### PR DESCRIPTION
**Pull request by Glorf**
## Summary by CodeRabbit

* **New Features**
  * Added sandbox backend supporting AWS ECS Fargate infrastructure with session management.
  * Enabled interactive command execution and file upload capabilities within sandboxes.
  * Integrated AWS credential validation and comprehensive error handling.

* **Documentation**
  * Added API documentation for the sandbox module and ECS Fargate usage.

* **Tests**
  * Added unit tests for sandbox module behavior and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by Glorf** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Evaluator/pull/624/commits/ad4fd4c33e725c5e94768806fe069b645c2043d1

**Comment by Glorf** [comment]
@coderabbitai full review

**Comment by Glorf** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Evaluator/pull/624/commits/216cff66ce86e34907a4d24f8f6c7f37abca19f6

File: docs/references/api/nemo-evaluator/sandbox/index.rst [hunk 1]
```diff
@@ -0,0 +1,55 @@
+``nemo_evaluator.sandbox``
```
Comment (agronskiy) [inline docs/references/api/nemo-evaluator/sandbox/index.rst:1]: nice! https://docs-nemo-evaluator.nvidia.com/mbien/sandbox/__delim__/references/api/nemo-evaluator/sandbox/index.html

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/base.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    def send_command(self, command: NemoSandboxCommand) -> None: ...
+
+    def capture_pane(self, capture_entire: bool = False) -> str: ...
+
+    def is_session_alive(self) -> bool: ...
+
+    def get_incremental_output(self) -> str: ...
+
+    def get_asciinema_timestamp(self) -> float: ...
```
Comment (agronskiy) [inline packages/nemo-evaluator/src/nemo_evaluator/sandbox/base.py:62]: `non-blocking`: funny to see `asciinema` which I knew as a demo capturing app here as a serious application for tracking the tty of the sandbox

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/base.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    def spin_up(
+        cls,
+        *,
+        task_id: str,
+        trial_name: str,
+        run_id: str,
+        pre_upload_paths: Iterable[Path] | None = None,
+        upload_dest_dir: str | None = None,
+        **kwargs,
+    ) -> ContextManager["NemoEvaluatorSandbox"]:
```
Comment (agronskiy) [inline packages/nemo-evaluator/src/nemo_evaluator/sandbox/base.py:91]: `nit`: more a question to Cursor I guess (or to the Cursor rules one uses): in presense of `from __future__ import annotation` one does need to take the own class's name into literal quotes.
image
image

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py [hunk 4]
```diff
@@ -0,0 +1,1172 @@
...
+
+class AwsCliMissingError(RuntimeError):
+    pass
+
+
+class EcsExecError(RuntimeError):
+    pass
+
+
+def _require_aws_sdks():
```
Comment (agronskiy) [inline packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py:85]: `thinking loud`: since it's not in pyproject 1) I find the return to try-catch dependency management suboptimal -- it would need additional steps from the user (`pre_cmd`) to install that into a docker if one forgets? 2) alternative: to add it as another subgroup to pyproject -- it better describes what the evaluator has -- and keep lazy loading (not `importlib` but normal `import blah` in this function)

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py [hunk 5]
```diff
@@ -0,0 +1,1172 @@
...
+            _as_text(getattr(cp, "stdout", ""))
+            + "\n"
+            + _as_text(getattr(cp, "stderr", ""))
+        ).splitlines()
+
+        filtered_lines: list[str] = []
+        for line in combined_lines:
+            if line.startswith("The Session Manager plugin was installed successfully"):
+                continue
+            if line.startswith("Starting session with SessionId:"):
```
Comment (agronskiy) [inline packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py]: `question`: is it intentional that one `startswith` has capitalized `SessionId` and the one below has lowercased -- does tmux plugin emit exactly that or it's a bug?

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py [hunk 6]
```diff
@@ -0,0 +1,1172 @@
...
+        self._s = sandbox
+        self.attrs = {"Config": {"User": ""}}
+
+    def exec_run(self, cmd: list[str], user: str = ""):
+        _ = user  # ECS Exec user switching is handled by remote shell / sudo if needed.
+        out = self._s._exec_capture(cmd=cmd, timeout_sec=180.0, check=True)
+        return type("ExecResult", (), {"exit_code": 0, "output": out.encode()})
+
+
+class EcsFargateTmuxSession(NemoSandboxSession):
```
Comment (agronskiy) [inline packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py:160]: `suggestion`: I would really add tmux version check and either emite a huge warning if we're beyond the current tmux version + 1 to be on the safe side. Since tmux `3.0` there was a good deal of development every minor and I see below that cursor relied on parsing number of emitted things that might break any time. Minor version updates of tmux don't guarantee backward compat of the emitted logs and magic strings

**Review by agronskiy — APPROVED** [review]
Amzaing, some questions and nits but unblocking. General request: ask cursor to add small (here, it does not pay off to have detailed ones) docstring to each tricky function just becasue it will make it simpler for human to skim. Or even a simple ascii diagram.

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py (hunk 4 above)
Comment (Glorf) [reply packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py:85]: That's a tricky one - I can add awscli to pyproject, yet sessionmanager needs to be handled separately in a dockerfile anyways :( : https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-working-with-install-plugin.html. For now it seemed like the best compromise we have

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/base.py (hunk 2 above)
Comment (Glorf) [reply packages/nemo-evaluator/src/nemo_evaluator/sandbox/base.py:62]: +1, research mode on :D

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py (hunk 6 above)
Comment (Glorf) [reply packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py:160]: I'm not super familiar with tmux development, do you have an intuition which version should we stick to? These will also depend on what tmux version the container provides so I'm a little uneasy with forcing a specific version if this tooling is expected to be generic

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py (hunk 5 above)
Comment (Glorf) [reply packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py]: Thanks for finding! Fixed now (it's lowercased so we don't make it case-sensitive)

File: packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py (hunk 6 above)
Comment (agronskiy) [reply packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py:160]: I would just research which one is working in our env and add a check on `tmux version`, issuing a warning into logs that the guaranteed one is blah and blah.

Comment (Glorf) [reply packages/nemo-evaluator/src/nemo_evaluator/sandbox/ecs_fargate.py:160]: added!

**Comment by Glorf** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Evaluator/pull/624/commits/1e31705d66568336d5382a949e5539a33ada057d

**Comment by Glorf** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Evaluator/pull/624/commits/efae04744279e1fec7f7123a0113d1297f771d74
